apps/keyboard/js/keyboard/input_method_manager.js

```diff
@@ -332,7 +332,16 @@ InputMethodLoader.prototype.getInputMethodAsync = function(imEngineName) {
   var p = new Promise(function(resolve, reject) {
     var script = document.createElement('script');
     script.onload = function() {
-      this.initInputMethod(imEngineName);
+      var initPromise = this.initInputMethod(imEngineName);
+
+      // Check if IME init function is Promise, if so, init should
+      // be resolved before activatation.
+      if (initPromise instanceof Promise) {
+        return initPromise.then(function() {
+          resolve(this._initializedIMEngines[imEngineName]);
+        }.bind(this));
+      }
```

Comment:
You should be able to avoid arrow pattern with proper usage promise. You can initialize the IMEngine in the `then()` block after the loading `new Promise()` block.